### PR TITLE
Output Link to ticket in found address CSV

### PR DESCRIPTION
**Pull request by gbirke**
The output CSV can now contain a ticket ID.
Move CSV generation to output class.
Delete CSVSourceDataReader - exported CSV has no ticket id and we're
using the database now.

### CODE REVIEW

File: src/UseCases/ExtractAddress/CSVFormatter.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?php
+declare( strict_types = 1 );
+
+namespace WMDE\OtrsExtractAddress\UseCases\ExtractAddress;
+
+trait CSVFormatter {
+
+	protected function formatAsCSV( array $row ): string {
```
Comment (JeroenDeDauw) [inline src/UseCases/ExtractAddress/CSVFormatter.php]: Why not private?

File: src/UseCases/ExtractAddress/CSVFormatter.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?php
+declare( strict_types = 1 );
+
+namespace WMDE\OtrsExtractAddress\UseCases\ExtractAddress;
+
+trait CSVFormatter {
```
Comment (JeroenDeDauw) [inline src/UseCases/ExtractAddress/CSVFormatter.php]: Why trait and not composition? If you don't want to inject the thing I'd still make a class and just instantiate it where it is used.

File: src/UseCases/ExtractAddress/CSVFormatter.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+	protected function formatAsCSV( array $row ): string {
+		return implode( ';', array_map( function( $v ) {
+			if ( $v ) {
+				return '"'. str_replace( '"', '""', $v ) . '"';
+			} else {
+				return '';
```
Comment (JeroenDeDauw) [inline src/UseCases/ExtractAddress/CSVFormatter.php]: Oh noez! an `if-esle`. You can use a guard followed by a return

File: src/UseCases/ExtractAddress/RejectedAddressWriter.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * @license GNU GPL v2+
+ * @author Gabriel Birke < gabriel.birke@wikimedia.de >
+ */
+class RejectedAddressWriter {
+
+	use CSVFormatter;
+
+	private $outputStream;
+
+	public function __construct( $outputStream ) {
```
Comment (JeroenDeDauw) [inline src/UseCases/ExtractAddress/RejectedAddressWriter.php]: Is it still not possible to type hint this in PHP 7.1?

**Review by JeroenDeDauw — COMMENTED** [review]
This talk is really boring... so you can has some reviews :)

**Comment by JeroenDeDauw** [comment]
+0.82

File: src/UseCases/ExtractAddress/RejectedAddressWriter.php (hunk 4 above)
Comment (gbirke) [reply src/UseCases/ExtractAddress/RejectedAddressWriter.php]: Nope, resources (DB resource, file handler resource), can't be type hinted yet. But I made another commit where I replace the resource with `SplFileObject`.

File: src/UseCases/ExtractAddress/RejectedAddressWriter.php [hunk 5]
```diff
@@ -14,14 +14,14 @@ class RejectedAddressWriter {
...
-	public function __construct( $outputStream ) {
+	public function __construct( \SplFileObject $outputStream ) {
 		$this->outputStream = $outputStream;
 		$this->csvFormatter = new CSVFormatter();
 	}
 
 	public function writeRow( SourceData $data, string $errorMsg ) {
 		// TODO add option to skip "not found" messages and export only other messages
-		fwrite( $this->outputStream, $this->csvFormatter->formatAsCSV( [
+		$this->outputStream->fwrite( $this->csvFormatter->formatAsCSV( [
```
Comment (JeroenDeDauw) [inline src/UseCases/ExtractAddress/RejectedAddressWriter.php]: cool... no more global call

**Comment by JeroenDeDauw** [comment]
I don't see any obvious issues but cannot +2 without understanding the tasks and tools involved here more. Presumably @KaiNissen is better up to speed?

**Pull request merged**